### PR TITLE
View App's recent changes

### PR DESCRIPTION
**Pull request by benrudolph**
@dimagi/ux This allows for users to view their most recent changes before deploying in a human readable way. I got so tired of guessing as to what I had changed. This isn't completely foolproof yet, but a decent start. It doesn't take into account settings or case changes. Also one cool thing about this is that it's complete agnostic to where it lives, so we can put this anywhere very easily. All we need is two app ids and selector to place the diff

screen shot 2017-03-15 at 2 34 57 pm

I also had to get extremely creative when trying to display the diff in a reasonable manner

- [x] Write JS tests
- [x] Check edge case for when there are no builds
- [x] Better styling
- [x] Documentation
- [x] DOMPurify to prevent XSS

cc: @biyeun

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        };
+    };
+
+    var QuestionDatum = function(json) {
+        var self = this;
+        this.comment = json.comment;
+        this.group = json.group;
+        this.hashtagValue = json.hashtagValue;
+        this.label = json.label;
+        this.options = json.optinos;
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: typo

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        this.required = json.required;
+        this.response = json.response;
+        this.tag = json.tag;
+        this.translations = json.translations;
+        this.type = json.type;
+        this.value = json.value;
+
+        this.toString = function() {
+            var lines =[
+                HtmlUtils.makeLi(self.label || '[unknown]', 'diff-question'),
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: translate `[unknown]`?

**Comment by orangejenny** [comment]
This is cool!

Two minor points/questions:
- What does it look like when there aren't any changes?
- Might want to make it clearer that this is changes to questions, not to settings, etc.

**Comment by benrudolph** [comment]
Good points, right now it just shows everything in black text if there are no changes. I'm still trying to figure out how to be clever about only showing what you need to see and hiding the rest while still giving enough context to know what the change is about.

Agree I should update title to Form Changes until it works for other things

**Comment by orangejenny** [comment]
> I'm still trying to figure out how to be clever about only showing what you need to see and hiding the rest while still giving enough context to know what the change is about.

Just musing, but I'd think you could hide any questions/forms/modules that didn't have anything inside them change. Although one way that could get confusing is if questions in a form are re-ordered...but then again, I'd guess that most form changes are about adding/removing questions or changing question properties and don't need a lot of context beyond identifying the module+form+question.

**Comment by orangejenny** [comment]
This would also be a cool supplement to the app diff page.

**Comment by benrudolph** [comment]
good idea i think that being the basic heuristic will work well to start

**Comment by emord** [comment]
Nice! I've always wanted something similar to this in the form builder's overwrite warning popover, but we don't have any app's json to compare so we need to interpret the xml direclty. Wonder if we could figure out someway to use this and make a temporary build for use in the form builder

**Comment by orangejenny** [comment]
I wonder if there's some kind of ellipsis-type indictor that could replace unchanged entities. The modal is going to get *really* long.

(If not obvious, this isn't feedback for this particular PR, just thinking about the future.)

**Comment by benrudolph** [comment]
Been thinking about it a little and the no changes detected thing isn't that hard. I think it'd make sense to make it expandable in the future just in case (similar to github). i agree that it will get very long. i'm just imagining this for the TB app..

@emord nice i discussed your idea briefly with ali and ahmad and both of them thought it was be very useful. i think it could be extended by accepting an optional xmlns param to limit which forms get returned

screen shot 2017-03-15 at 11 04 12 pm

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        this.id = json.id;
+        this.name = json.name;
+        this.shortComment = json.short_comment || '';
+        this.forms = _.map(json.forms, function(form) { return new FormDatum(form); });
+
+        this.toString = function() {
+            var lines = [
+                // We want these to be considered one line
+                (
+                    HtmlUtils.makeLi(sanitize(self.name.en), 'diff-module', 'folder-open') +
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Expected indentation of 16 space characters but found 20. (indent)

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        this.id = json.id;
+        this.name = json.name;
+        this.shortComment = json.short_comment || '';
+        this.questions = _.map(json.questions, function(q) { return new QuestionDatum(q); });
+
+        this.toString = function() {
+            var lines = [
+                // We want these to be considered one line
+                (
+                    HtmlUtils.makeLi(sanitize(self.name.en), 'diff-form', 'file-o') +
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Expected indentation of 16 space characters but found 20. (indent)

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+            ];
+            return lines.join('\n');
+        };
+    };
+
+    /**
+     * Represents the question data structure and renders it to an HTML string
+     */
+    var QuestionDatum = function(json) {
+        console.log(json);
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Unexpected console statement. (no-console)

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        this.tag = json.tag;
+        this.translations = json.translations;
+        this.type = json.type;
+        this.value = json.value;
+
+        this.toString = function() {
+            var lines =[
+                // We want these to be considered one line
+                (
+                    HtmlUtils.makeLi(sanitize(self.label) || gettext('[unknown]'), 'diff-question') +
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Expected indentation of 16 space characters but found 20. (indent)

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        this.name = json.name;
+        this.options = options;
+        this.shortComment = json.short_comment || '';
+        this.forms = _.map(json.forms, function(form) { return new FormDatum(form, self.options); });
+
+        this.toString = function() {
+            var lines = [
+                // We want these to be considered one line
+                (
+                    HtmlUtils.makeLi(sanitize(self.name[self.options.lang]), 'diff-module', 'folder-open') +
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Expected indentation of 16 space characters but found 20. (indent)

File: corehq/apps/app_manager/static/app_manager/js/app_diff.js [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        this.name = json.name;
+        this.options = options;
+        this.shortComment = json.short_comment || '';
+        this.questions = _.map(json.questions, function(q) { return new QuestionDatum(q, self.options); });
+
+        this.toString = function() {
+            var lines = [
+                // We want these to be considered one line
+                (
+                    HtmlUtils.makeLi(sanitize(self.name[self.options.lang]), 'diff-form', 'file-o') +
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/app_diff.js]: Expected indentation of 16 space characters but found 20. (indent)

File: corehq/apps/app_manager/static/app_manager/js/releases.js [hunk 9]
```diff
@@ -193,6 +193,7 @@ hqDefine('app_manager/js/releases.js', function () {
     function ReleasesMain(o) {
         /* {fetchUrl, deleteUrl} */
         var AsyncDownloader = hqImport('app_manager/js/download_async_modal.js').AsyncDownloader;
+        var AppDiff = hqImport('app_manager/js/app_diff.js')
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/releases.js]: Missing semicolon. (semi)

**Comment by benrudolph** [comment]
Ok I think this is ready for re-review. I didn't make any assumption about how it work in app manager v2, so left that out for now. also added a button for view the changes between each deploy 😸

**Comment by snopoke** [comment]
Wow, this is cool.

**Comment by benrudolph** [comment]
✅

File: corehq/apps/app_manager/templates/app_manager/v1/partials/app_diff_modal.html [hunk 10]
```diff
@@ -0,0 +1,16 @@
+<div class="modal fade" id="app-diff-modal" tabindex="-1" role="dialog">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/app_diff_modal.html]: `trans` for this file?

**Comment by orangejenny** [comment]
Looks good to me.

> I didn't make any assumption about how it work in app manager v2, so left that out for now.

It'd be better to add this to v2, I don't think we're doing anything that would affect this. Don't want this to get forgotten about and then disappear when we release v2. Feel free to self-merge if you'd rather do that in another PR.

**Review by orangejenny — APPROVED** [review]

**Comment by benrudolph** [comment]
thanks for the review @orangejenny. i will follow up with that feedback in another PR

**Pull request merged**